### DISCUSSION TITLE
Discussion tracking: Event dispatching / listening between AFFINE and Blocksuite

### DISCUSSION THREAD

**Discussion by thorseraq** [toeverything/blocksuite#198] [Q&A] [answered]
AFFINE uses blocksuite as the EditorProvider. When reading the code of AFFINE (of the pathfinder branch), I met a question that when window dispatches an event, (the Capture phrase is disabled by default) this event bubbles up from the "window" element, which element will receive this event? (window seems don't have any parent element) . 
<img width="1384" alt="image" src="https://user-images.githubusercontent.com/20554850/200990045-3f1762ef-05e0-497a-b873-8ad4ebbd970f.png">

I referred [Bubbling and capturing explained](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#bubbling_and_capturing_explained). 

> In the bubbling phase, the exact opposite of the capturing phase occurs:
> The browser checks to see if **the direct parent of the clicked element** has a click event handler registered on it for the bubbling phase, and runs it if so.
> Then it moves on to the **next immediate ancestor element** and does the same thing, then the next one, and so on until it reaches the <html> element.

**Comment by thorseraq** [comment]
It seems that the `<editor-container></editor-container>` added an event listener to window, but "this" remains referring to the` <editor-container></editor-container> `itself in the event handler (utilized arrow function)
![image](https://user-images.githubusercontent.com/20554850/200993730-e364736b-157c-4c8a-a449-2395b36a26aa.png)

**Comment by thorseraq** [comment]
After discussions with main maintainers (@colelawrence @doodlewind  ), this turns out to be a workaround for now (As window exists in browser context, so components in whatever framework (Blocksuite uses Lit, AFFINE uses react) can add an event listener on it or dispatch an event on it. ), and @doodlewind suggests that **a new comprehensive API design about how to reuse EditorContainer is welcomed**.

**Comment by lawvs** [comment, marked as answer]
> I met a question that when window dispatches an event, (the Capture phrase is disabled by default) this event bubbles up from the "window" element, which element will receive this event? (window seems don't have any parent element) .

You can read the DOM spec for a more detailed explanation. The parent of the window is `null` and the event will stop bubbling.

https://dom.spec.whatwg.org/#dispatching-events

> Each [EventTarget](https://dom.spec.whatwg.org/#eventtarget) object also has an associated get the parent algorithm, which takes an [event](https://dom.spec.whatwg.org/#concept-event) event, and returns an [EventTarget](https://dom.spec.whatwg.org/#eventtarget) object. Unless specified otherwise **it returns null**.
>
> [Nodes](https://dom.spec.whatwg.org/#concept-node), [shadow roots](https://dom.spec.whatwg.org/#concept-shadow-root), and [documents](https://dom.spec.whatwg.org/#concept-document) override the [get the parent](https://dom.spec.whatwg.org/#get-the-parent) algorithm.
>
> https://dom.spec.whatwg.org/#get-the-parent

**Comment by thorseraq** [comment]
thanks, It's very detailed, covering even situation of Nodes, shadow roots (Occasionally understand more about event bubbling rules)
